gen/internals.py

```diff
@@ -0,0 +1,464 @@
+            raise CalculatorError(ex.message, ex.chain + ['while calculating {}'.format(name)]) from ex
+        except:
+            self._arguments[name] = None
+            raise
+        finally:
+            self.__in_progress.remove(name)
+
+    def _calculate_target(self, target):
+        def evaluate_var(name):
+            try:
+                self._get(name)
+            except CalculatorError as ex:
+                log.debug("Error calculating %s: %s. Chain: %s", name, ex.message, ex.chain)
+
+        for name in target.variables:
+            evaluate_var(name)
+
+        for name, sub_scope in target.sub_scopes.items():
+            if name not in self._arguments:
+                evaluate_var(name)
+
+            # If the internal arg is None, there was an error, don't check if it
+            # is a legal choice.
+            if self._arguments[name] is None:
+                continue
+
+            choice = self._get(name)
+
+            if choice not in sub_scope.cases:
+                self._errors[name] = "Invalid choice {}. Must choose one of {}".format(
+                    choice, ", ".join(sorted(sub_scope.keys())))
+                continue
+
+            self._calculate_target(sub_scope.cases[choice])
+
+        # Perform all multi-argument validations
+        for parameter_set, validate_fn in self._multi_arg_validate.items():
+            # Build up argument map for validate function. If any arguments are
+            # unset then skip this validate function.
+            kwargs = dict()
+            skip = False
+            for parameter in parameter_set:
+                if parameter not in self._arguments or self._arguments[parameter] is None:
+                    skip = True
+                    break
+                kwargs[parameter] = self._arguments[parameter]
+            if skip:
+                continue
+
+            # Call the validation function, catching AssertionErrors and turning them into errors in
+            # the error dictionary.
+            try:
+                validate_fn(**kwargs)
+            except AssertionError as ex:
+                self._errors[parameter_set] = ex.args[0]
+
+        # TODO(cmaloney): Return per-target results with the path to calculate each argument and the
+        # full set of arguments touched included.
+        return self._arguments
+
+    # Force calculation of all arguments by accessing the arguments in this
+    # scope and recursively all sub-scopes.
+    def calculate(self, targets):
+        for target in targets:
+            self._calculate_target(target)
+
+        if len(self._errors) or len(self._unset):
+            raise ValidationError(self._errors, self._unset)
+
+        return self._arguments
+
+
+def resolve_configuration(sources: list, targets: list, user_arguments: dict):
+
+    # Make sure all user provided arguments are strings.
+    # TODO(cmaloney): Loosen this restriction  / allow arbitrary types as long
```

Comment:
I don't believe generally making JIRAs for TODOs creates useful data / input points. We should at some point make a tool which gathers TODOs and presents them, and maybe gives an easy button to make a JIRA for them, but manually keeping TODOs in sync with JIRA is going to be a very long task without a lot of benefit.

It's also not the standard for the codebase at this point in time.